File: acme-admin/src/main/java/com/acme/admin/controller/TaxController.java
package com.acme.admin.controller;

import com.acme.admin.dto.TaxDtos;
import com.acme.admin.service.TaxService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/taxes")
public class TaxController {

    private final TaxService svc;

    public TaxController(TaxService svc) {
        this.svc = svc;
    }

    @GetMapping
    public ResponseEntity<List<TaxDtos.TaxResponse>> list() {
        return ResponseEntity.ok(svc.list());
    }

    @PostMapping
    public ResponseEntity<TaxDtos.TaxResponse> create(@RequestBody TaxDtos.CreateTaxRequest req) {
        return ResponseEntity.status(HttpStatus.CREATED).body(svc.create(req));
    }

    @PutMapping("/{id}")
    public ResponseEntity<TaxDtos.TaxResponse> update(
        @PathVariable("id") UUID id,
        @RequestBody TaxDtos.UpdateTaxRequest req
    ) {
        return ResponseEntity.ok(svc.update(id, req));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable("id") UUID id) {
        svc.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: acme-admin-dotnet/Controllers/TaxController.cs
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Acme.Admin.Api.Validation;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/taxes")]
public sealed class TaxController(TaxService service) : ControllerBase
{
    [HttpGet]
    [Authorize(Policy = PermissionPolicies.TaxesRead)]
    public async Task<ActionResult<PagedResult<TaxDtos.TaxResponse>>> List(
        [FromQuery] int page = Paging.DefaultPage,
        [FromQuery] int size = Paging.DefaultSize,
        CancellationToken ct = default)
    {
        var (p, s) = Paging.Normalize(page, size);
        return Ok(await service.ListAsync(p, s, ct));
    }

    [HttpPost]
    [Authorize(Policy = PermissionPolicies.TaxesWrite)]
    public async Task<ActionResult<TaxDtos.TaxResponse>> Create([FromBody] TaxDtos.CreateTaxRequest req, CancellationToken ct)
    {
        var created = await service.CreateAsync(req, ct);
        return StatusCode(StatusCodes.Status201Created, created);
    }

    [HttpPut("{id:guid}")]
    [Authorize(Policy = PermissionPolicies.TaxesWrite)]
    public async Task<ActionResult<TaxDtos.TaxResponse>> Update([FromRoute] Guid id, [FromBody] TaxDtos.UpdateTaxRequest req, CancellationToken ct)
    {
        return Ok(await service.UpdateAsync(id, req, ct));
    }

    [HttpDelete("{id:guid}")]
    [Authorize(Policy = PermissionPolicies.TaxesWrite)]
    public async Task<IActionResult> Delete([FromRoute] Guid id, CancellationToken ct)
    {
        await service.DeleteAsync(id, ct);
        return NoContent();
    }
}
